### DISCUSSION TITLE
r2r sink

### DISCUSSION THREAD

**Discussion by parabyte** [pschatzmann/arduino-audio-tools#1675] [Q&A] [answered]
is there any any examples on how to use audiotools with a crude 8bit r2r dac on a arduino on a stm32?

**Comment by pschatzmann** [comment, marked as answer]
No, I think the quality would be too bad.
You can try to use the internal Dac or Pwm instead.

What Micocontroller do you plan to use?

**Comment by parabyte** [comment]
stm32 blackpill F4 variant, i got your snapclient working on it with ethernet,

**Comment by pschatzmann** [comment]
As far as I know, the blackpill does not have any DAC, but you can [still use PWM](https://www.pschatzmann.ch/home/2022/10/03/arduino-stm32-audio-pwm-output/) which should give a better audio quality then your 8 step r2r.

ps. I2S should work as well...

**Comment by parabyte** [comment]
**sadly    the above  code as is throws this error.**


/tmp/.arduinoIDE-unsaved2024723-28154-1voj54y.090d/sketch_aug23f/sketch_aug23f.ino:7:1: error: 'PWMAudioStream' does not name a type
    7 |   // put your main code here, to run repeatedly:
      | ^ ~~~~~~~~~~~~
/tmp/.arduinoIDE-unsaved2024723-28154-1voj54y.090d/sketch_aug23f/sketch_aug23f.ino:8:19: error: 'pwm' was not declared in this scope
    8 | 
      |                   ^  
/tmp/.arduinoIDE-unsaved2024723-28154-1voj54y.090d/sketch_aug23f/sketch_aug23f.ino: In function 'void setup()':
/tmp/.arduinoIDE-unsaved2024723-28154-1voj54y.090d/sketch_aug23f/sketch_aug23f.ino:18:17: error: 'pwm' was not declared in this scope

exit status 1

Compilation error: 'PWMAudioStream' does not name a type

**Comment by parabyte** [comment]
any chance of any pointers to do this r2r dac thing even though its a bad idea, i spent a day populating a vero board with resistors with my daughter and just want to show it off doing something useful, :)

**Comment by pschatzmann** [comment]
This should be quite easy to implement: [Define your own output class](https://github.com/pschatzmann/arduino-audio-tools/wiki/Writing-your-Custom-Input&Output-Class) or use a [CallbackStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_callback_stream.html) and define the write callback.

Write the data into a buffer and output the data with an AudioTimer: There you just map the input values to your pins.


If your input provides stereo data you need to [convert it to mono first](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format) or you need to use a second ladder.

I can't reproduce your issue with the PWMAudioStream: what Board Part Number did you select ? I tried with Blackpill401CC, Blackpill401CE and Blackpill411CE. For some I needed to uncomment the USE_I2S in AudioConfig.h

**Comment by pschatzmann** [comment]
I have added a [DRAFT implementation](https://github.com/pschatzmann/arduino-audio-tools/commit/734b4354b8360871ea3666f2d5e75294e0111423): I let you test and debug it...

**Comment by parabyte** [comment]
**i seem to be hitting this error** 

/home/gatekeeper/Arduino/libraries/stm32-i2s/src/stm32-i2s.h:131:20: error: 'STM_I2S_PINS' was not declared in this scope
  131 |   I2SPin pins[5] = STM_I2S_PINS;
      |                    ^~~~~~~~~~~~
/home/gatekeeper/Arduino/libraries/stm32-i2s/src/stm32-i2s.h: In member function 'void stm32_i2s::Stm32I2sClass::MX_I2S3_Init()':
/home/gatekeeper/Arduino/libraries/stm32-i2s/src/stm32-i2s.h:481:22: error: 'SPI_INSTANCE_FOR_I2S' was not declared in this scope
  481 |     hi2s3.Instance = SPI_INSTANCE_FOR_I2S;

**Comment by parabyte** [comment]
its building fine if i disable i2s following the discussion we had https://github.com/pschatzmann/arduino-snapclient/discussions/34

**Comment by parabyte** [comment]
an example of how you would do r2r in a sketch would be fantastic

**Comment by pschatzmann** [comment]
It's basically like any other output class: The only special thing you need to do is to define the channel1_pins in the configuration for mono or channel1_pins and channel2_pins for stereo.

It's up to you how many pins you want to use...

```C++
  // start 
  auto config = out.defaultConfig();
  config.copyFrom(info); 
  config.channel1_pins = {10,11,12,13,14,15,16,17};
  if (!out.begin(config)){
    Serial.println("error");
    while(true);
  }

```

I am not sure about the max sample rate that will work since this is limited by the speed of digitalWrite()

**Comment by parabyte** [comment]
i cant seem to get sound output from the snapclient into this new lovely lib you written, ?

**Comment by pschatzmann** [comment]
You need to do some proper testing with some simple test sketches first:
1. check the mapping of the input values (range for 16 bits: −32,768 to +32,767) to output values (range for 8 bits: 0—255) 
2. check the activated pins for all output values
3. check the generated voltage for the full output value range

After you confirmed that the voltages are right you can start with a sine test case and a min sample rate of 8000 and increase the sample rate to determine the max supported rate

**Reply by parabyte** [reply]
just ordered a oscilloscope so i can give you some meaningful measurements

**Comment by parabyte** [comment]
iv just crudely measured with my multimeter. and all seems well, from what i can hear and measure crudely with my multimeter this is producing tones and working fine with the various tests you suggested.

**Comment by pschatzmann** [comment]
So, if you connect the DAC to some earphones and use the sine test output sketch;
What is the max sample rate that still works ?

**Comment by parabyte** [comment]
on my particular blackpill board its maxing out at 10000 unless i am
mistaken

**Reply by parabyte** [reply]
mine is cc variant

**Comment by parabyte** [comment]
audio-tools now has a r2r driver!

thanks!
